aten/src/ATen/native/mps/operations/Randperm.mm

```diff
@@ -0,0 +1,80 @@
+//  Copyright © 2022 Apple Inc.
+
+#include <ATen/native/mps/OperationUtils.h>
+#include <ATen/native/TensorFactories.h>
+
+namespace at {
+namespace native {
+
+Tensor& randperm_out_mps(int64_t n, c10::optional<Generator> generator, Tensor& result) {
+  if (!MPSDevice::getInstance()->macOS_13_0()) {
+    TORCH_WARN_ONCE("MPS: randperm op is supported natively starting from macOS 13.0. ",
+                    "Falling back on CPU. This may have performace implications.");
+
+    result = result.to("cpu");
+    result = at::randperm_out(result, n).to("mps");
+    return result;
+  }
+
+  using namespace mps;
+  TORCH_CHECK(n >= 0, "n must be non-negative, got", n);
+  TORCH_CHECK(!generator.has_value() || (generator.has_value() && result.device() == generator->device()), "Expected a '", result.device(), "' generator device but found '", generator->device(), "'");
+  check_supported_max_int_with_precision(n, result);
+  result.resize_({n});
+
+  auto uniform_mps = at::native::empty_mps(result.sizes(), kFloat, c10::nullopt, kMPS);
+  uniform_mps = at::uniform(uniform_mps, 0.0, 1.0, generator);
+
+  MPSStream *stream = getCurrentMPSStream();
+
+  struct CachedGraph : public MPSCachedGraph
+  {
+    CachedGraph(MPSGraph *graph) : MPSCachedGraph(graph) {}
+    MPSGraphTensor* uniformTensor_ = nil;
+    MPSGraphTensor* outputTensor_ = nil;
+  };
+
+  MPSGraphCache* cache_ = MPSGraphCache::getInstance();
+  @autoreleasepool {
+    string key = "randperm_out_mps" + getTensorsStringKey({result}) ;
+    CachedGraph* cachedGraph = static_cast<CachedGraph *>(cache_->LookUp(key));
+
+    if(!cachedGraph) {
+      MPSCachedGraph *tmpCachedGraph = cache_->CreateCachedGraph(key, ^ MPSCachedGraph * () {
+
+        CachedGraph *newCachedGraph = nil;
+        @autoreleasepool {
+
+          MPSGraph* mpsGraph = make_mps_graph();
+          newCachedGraph = new CachedGraph(mpsGraph);
+
+          newCachedGraph->uniformTensor_ = mpsGraphRankedPlaceHolder(mpsGraph, uniform_mps);
+          newCachedGraph->outputTensor_ = [mpsGraph argSortWithTensor:newCachedGraph->uniformTensor_
```

Comment:
If you move this to Distribution.mm, you could use the common `random_mps_impl()`, and just add `argSortWithTensor` to the RandomOpBlock, like this:
```
...
  mps::RandomOpBlock random_op_block = ^RandomOpFn(cachedGraph, randomTensor) {
    MPSGraph* mpsGraph = cachedGraph->graph();
   return [mpsGraph argSortWithTensor:randomTensor ...]
  };
  return mps::random_mps_impl<double>(self, 0.0, 1.0, c10::nullopt, c10::nullopt,
                                      MPSGraphRandomDistributionUniform, gen,
                                      "randperm_mps_", random_op_block);
...
```
See implementation of `exponential_mps_()` for an example. Also, this way you'll do runMPSGraph once only.

Reply:
Thanks - that code block seems really useful. I'll move this code there

Reply:
But in this case I'd have two tensors to pass in the code block - `self` and `result`, and the code block accepts only 1.

Reply:
Thanks- I've updated it